crates/ibc/src/core/ics04_channel/context.rs

```diff
@@ -127,10 +127,9 @@ pub trait ChannelReader {
 
     /// Returns the current timestamp of the local chain.
     fn host_timestamp(&self) -> Result<Timestamp, ChannelError> {
-        let pending_consensus_state = self
-            .pending_host_consensus_state()
-            .expect("host must have pending consensus state");
-        Ok(pending_consensus_state.timestamp())
+        let host_height = self.host_height()?;
+        let host_consensus_state = self.host_consensus_state(&host_height)?;
+        Ok(host_consensus_state.timestamp())
```

Comment:
I don't think we can assume that this will work for all hosts. Basecoin-rs is an example: we added `pending_host_consensus_state()` because `host_consensus_state()` was not quite right. I think it should [simply be made abstract](https://github.com/cosmos/ibc-rs/issues/418#issuecomment-1426200776); each host can implement it however they see fit. We should do it in a separate PR though to keep PRs as self-contained as possible.

Reply:
Actually, moving the `host_timestamp()` methods brought a need to handle them here. but, sure alright.

Reply:
what do you mean by "moving the `host_timestamp()` methods brought a need to handle them here"?